ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp

```diff
@@ -0,0 +1,243 @@
+                    T x2 = rois[rois_transform.index({roi_index, 2})] * spatial_scale;
+                    T y2 = rois[rois_transform.index({roi_index, 3})] * spatial_scale;
+
+                    T roi_width = std::max(x2 - x1, static_cast<T>(1.0));
+                    T roi_height = std::max(y2 - y1, static_cast<T>(1.0));
+
+                    T bin_width = roi_width / pooled_width;
+                    T bin_height = roi_height / pooled_height;
+
+                    uint64_t num_samples_in_bin = sampling_ratio * sampling_ratio;
+
+                    T sample_distance_x = bin_width / static_cast<T>(sampling_ratio);
+                    T sample_distance_y = bin_height / static_cast<T>(sampling_ratio);
+
+                    std::vector<std::pair<uint64_t, uint64_t>> pooling_points;
+                    std::vector<T> pooling_weights;
+
+                    pooling_points.reserve(4 * num_samples_in_bin * pooled_height * pooled_width);
+                    pooling_weights.reserve(4 * num_samples_in_bin * pooled_height * pooled_width);
+
+                    // Save the sample coords and weights as they will be identical across all
+                    // channels
+                    for (int64_t y_bin_ind = 0; y_bin_ind < pooled_height; y_bin_ind++)
+                    {
+                        for (int64_t x_bin_ind = 0; x_bin_ind < pooled_width; x_bin_ind++)
+                        {
+                            for (int64_t y_sample_ind = 0; y_sample_ind < sampling_ratio;
+                                 y_sample_ind++)
+                            {
+                                T sample_y = y1 + static_cast<T>(y_bin_ind) * bin_height +
+                                             sample_distance_y * (static_cast<T>(y_sample_ind) +
+                                                                  static_cast<T>(0.5f));
+
+                                for (int64_t x_sample_ind = 0; x_sample_ind < sampling_ratio;
+                                     x_sample_ind++)
+                                {
+                                    T sample_x = x1 + static_cast<T>(x_bin_ind) * bin_width +
+                                                 sample_distance_x * (static_cast<T>(x_sample_ind) +
+                                                                      static_cast<T>(0.5f));
+
+                                    if (sample_x < -1.0 || sample_x > feature_map_width ||
+                                        sample_y < -1.0 || sample_y > feature_map_height)
+                                    {
+                                        // For this sample we save 4x point (0,0) with weight 0
+                                        pooling_points.push_back({0, 0});
+                                        pooling_points.push_back({0, 0});
+                                        pooling_points.push_back({0, 0});
+                                        pooling_points.push_back({0, 0});
```

Comment:
I recommend using the insert method instead https://en.cppreference.com/w/cpp/container/vector/insert 

`pooling_points.insert(pooling_points.end(), 4, {0, 0})`

Reply:
I did change it here, left the push_backs that occur later, because i`m worried about decreasing readability when the inserted values are being calculated